src/middlewared/middlewared/plugins/vm.py

```diff
@@ -810,39 +811,85 @@ def activate_sharefs(self, pool_name=None):
-        # Get the user memory and keep space for ARC.
-        throttled_user_mem = int(usermem[0].value - throttled_arc_max)
-        # Potential memory used by guests.
-        memory_used = guest_mem_used['RPRD'] + guest_mem_used['RNP']
-
-        vms_memory = memory_used + memory
-        if vms_memory <= throttled_user_mem:
-            if max_arc[0].value > throttled_arc_max:
-                if max(max_arc[0].value - memory, 0) != 0:
-                    self.logger.info('===> Setting ARC FROM: {} TO: {}'.format(max_arc[0].value, max_arc[0].value - memory))
-                    sysctl.filter('vfs.zfs.arc_max')[0].value = max_arc[0].value - memory
-            return True
-        else:
+    @accepts(Bool('overcommit', default=False))
+    def get_available_memory(self, overcommit):
+        """
+        Get the current maximum amount of available memory to be allocated for VMs.
+
+        If `overcommit` is true only the current used memory of running VMs will be accounted for.
+        If false all memory (including unused) of runnings VMs will be accounted for.
+
+        This will include memory shrinking ZFS ARC to the minimum.
+
+        Memory is of course a very "volatile" resource, values may change abruptly between a
+        second but I deem it good enough to give the user a clue about how much memory is
+        available at the current moment and if a VM should be allowed to be launched.
+        """
+        # Use 90% of free memory to play safe
+        free = int(psutil.virtual_memory().free * 0.9)
+
+        # swap used space is accounted for used physical memory because
+        # 1. processes (including VMs) can be swapped out
+        # 2. we want to avoid using swap
+        swap_used = psutil.swap_memory().used * sysctl.filter('hw.pagesize')[0].value
+
+        # Difference between current ARC total size and the minimum allowed
+        arc_total = sysctl.filter('kstat.zfs.misc.arcstats.size')[0].value
+        arc_min = sysctl.filter('vfs.zfs.arc_min')[0].value
+        arc_shrink = max(0, arc_total - arc_min)
+
+        vms_memory_used = 0
+        if overcommit is False:
+            # If overcommit is not wanted its verified how much physical memory
+            # the bhyve process is currently using and add the maximum memory its
+            # supposed to have.
+            for vm in self.middleware.call_sync('vm.query'):
+                status = self.middleware.call_sync('vm.status', vm['id'])
+                if status['pid']:
+                    try:
+                        p = psutil.Process(status['pid'])
+                    except psutil.NoSuchProcess:
+                        continue
+                    memory_info = p.memory_info()._asdict()
+                    memory_info.pop('vms')
+                    vms_memory_used += (vm['memory'] * 1024 * 1024) - sum(memory_info.values())
+
+        return max(0, free + arc_shrink - vms_memory_used - swap_used)
```

Comment:
Amount of swap used is not included in `psutil.virtual_memory` output, why are we subtracting it here?

Reply:
Its explained above.